src/main/capi/datetime-c.cpp

```diff
@@ -99,3 +103,105 @@ duckdb_timestamp duckdb_to_timestamp(duckdb_timestamp_struct ts) {
 bool duckdb_is_finite_timestamp(duckdb_timestamp ts) {
 	return Timestamp::IsFinite(timestamp_t(ts.micros));
 }
+
+duckdb_timestamp_struct duckdb_from_timestamp_s(duckdb_timestamp_s ts_s) {
+	date_t date;
+	dtime_t time;
+	timestamp_t ts_us = Timestamp::FromEpochSecondsPossiblyInfinite(ts_s.seconds);
+	Timestamp::Convert(ts_us, date, time);
+
+	duckdb_date ddate;
+	ddate.days = date.days;
+
+	duckdb_time dtime;
+	dtime.micros = time.micros;
+
+	duckdb_timestamp_struct result;
+	result.date = duckdb_from_date(ddate);
+	result.time = duckdb_from_time(dtime);
+	return result;
+}
+
+duckdb_timestamp_s duckdb_to_timestamp_s(duckdb_timestamp_struct ts) {
+	date_t date = date_t(duckdb_to_date(ts.date).days);
+	dtime_t time = dtime_t(duckdb_to_time(ts.time).micros);
+
+	duckdb_timestamp_s result;
+	result.seconds = Timestamp::GetEpochSeconds(Timestamp::FromDatetime(date, time));
+	return result;
+}
+
+bool duckdb_is_finite_timestamp_s(duckdb_timestamp_s ts_s) {
+	return Timestamp::IsFinite(timestamp_t(ts_s.seconds));
+}
+
+duckdb_timestamp_struct duckdb_from_timestamp_ms(duckdb_timestamp_ms ts_ms) {
+	date_t date;
+	dtime_t time;
+	timestamp_t ts_us = Timestamp::FromEpochMsPossiblyInfinite(ts_ms.millis);
+	Timestamp::Convert(ts_us, date, time);
+
+	duckdb_date ddate;
+	ddate.days = date.days;
+
+	duckdb_time dtime;
+	dtime.micros = time.micros;
+
+	duckdb_timestamp_struct result;
+	result.date = duckdb_from_date(ddate);
+	result.time = duckdb_from_time(dtime);
+	return result;
+}
+
+duckdb_timestamp_ms duckdb_to_timestamp_ms(duckdb_timestamp_struct ts) {
+	date_t date = date_t(duckdb_to_date(ts.date).days);
+	dtime_t time = dtime_t(duckdb_to_time(ts.time).micros);
+
+	duckdb_timestamp_ms result;
+	result.millis = Timestamp::GetEpochMs(Timestamp::FromDatetime(date, time));
+	return result;
+}
+
+bool duckdb_is_finite_timestamp_ms(duckdb_timestamp_ms ts_ms) {
+	return Timestamp::IsFinite(timestamp_t(ts_ms.millis));
+}
+
+duckdb_timestamp_ns_struct duckdb_from_timestamp_ns(duckdb_timestamp_ns ts_ns) {
+	date_t date;
+	dtime_t time;
+	int32_t nanos;
+	timestamp_ns_t ts_ns_t;
+	ts_ns_t.value = ts_ns.nanos;
+	Timestamp::Convert(ts_ns_t, date, time, nanos);
+
+	duckdb_date ddate;
+	ddate.days = date.days;
+
+	duckdb_time dtime;
+	dtime.micros = time.micros;
+
+	duckdb_timestamp_ns_struct result;
+	result.date = duckdb_from_date(ddate);
+	result.time = duckdb_from_time(dtime);
+	result.nanos = nanos;
+	return result;
+}
+
+duckdb_timestamp_ns duckdb_to_timestamp_ns(duckdb_timestamp_ns_struct ts) {
+	date_t date = date_t(duckdb_to_date(ts.date).days);
+	dtime_t time = dtime_t(duckdb_to_time(ts.time).micros);
+	int32_t nanos = ts.nanos;
+
+	timestamp_ns_t ts_ns;
+	if (!Timestamp::TryFromTimestampNanos(Timestamp::FromDatetime(date, time), nanos, ts_ns)) {
+		throw duckdb::ConversionException("Overflow exception in date/time -> timestamp conversion");
+	}
```

Comment:
Hm, a lot of the conversion functions user here throw, including in existing functions (e.g. `Timestamp::Convert` used by `duckdb_from_timestamp`, `Timestamp::FromDatetime` used by `duckdb_to_timestamp`), so this affects more than just this function.

Many other functions in the C API return a `duckdb_state` to indicate success or error, and have a separate function to get the error string on the relevant object. But there isn't a relevant object in this case.

Looks like the `out_error` pattern has one precedent: `duckdb_open_ext`. I could follow that for the new functions I'm adding.

But fixing all such problems here would require making breaking changes to existing functions, or more likely adding new ones and deprecating the old ones. That's out of scope for this change.

Reply:
It seems a more common pattern is to catch any errors returned by the C++ layer and just return DuckDBError or DuckDBSuccess, rather than try to transfer the error message. The error messages aren't especially useful for these conversion messages; they're all some variation on "overflow" or "out of range", or just "could not convert".

This also avoids the problem of memory management for `out_error`. The one place it is used (`duckdb_open_ext`) expects the caller to free it. That would be pretty awkward for the caller to have to worry about for these conversion functions.

Reply:
I've updated the NS conversion functions to have an out parameter for the result and return a duckdb_state to indicate success or failure. Let me know if this looks like a good pattern, and I'll update the functions for the other types accordingly.

Reply:
> It seems a more common pattern is to catch any errors returned by the C++ layer and just return DuckDBError or DuckDBSuccess, rather than try to transfer the error message

Yes, in many places, we also extract the error message and store it in the wrapper structure (e.g., the appender wrapper). But I don't think we want to do that here.

> I've updated the NS conversion functions to have an out parameter for the result and return a duckdb_state to indicate success or failure. Let me know if this looks like a good pattern, and I'll update the functions for the other types accordingly - or try some other approach.

Yes, great. I think this is lightweight enough (no deallocations on error and such), and at the same time, there is a way to indicate a failure.